File: client/auditorclient/record.py
"""Records"""

from __future__ import annotations  # not necessary in 3.10
import json
from .errors import InsufficientParametersError


class Components:
    def __init__(self):
        self._components = []

    def __str__(self) -> str:
        return self._components.__str__()

    def __eq__(self, other: Components) -> bool:
        return self._components == other._components

    def add_component(self, name: str, amount: int, factor: float) -> Components:
        self._components.append({"name": name, "amount": amount, "factor": factor})
        return self

    def get(self) -> [dict]:
        return self._components


class Record:
    def __init__(
        self,
        record_id: str = None,
        site_id: str = None,
        user_id: str = None,
        group_id: str = None,
        components: Components = None,
        json_str: str = None,
    ):
        if (
            record_id is None
            or site_id is None
            or user_id is None
            or group_id is None
            or components is None
        ) and json_str is None:
            raise InsufficientParametersError
        if json_str is None:
            self._record_id = record_id
            self._site_id = site_id
            self._user_id = user_id
            self._group_id = group_id
            self._components = components
            self._start_time = None
            self._stop_time = None
            #  self._runtime = None
        else:
            d = json.loads(json_str)
            c = Components()
            for comp in d["components"]:
                c.add_component(comp["name"], comp["amount"], comp["factor"])
            self._record_id = d["record_id"]
            self._site_id = d["site_id"]
            self._user_id = d["user_id"]
            self._group_id = d["group_id"]
            self._components = c
            self._start_time = d["start_time"]
            self._stop_time = d["stop_time"]
            #  self._runtime = d["runtime"]

    def __str__(self) -> str:
        return self.as_dict().__str__()

    def __eq__(self, other: Record) -> bool:
        return (
            self._record_id == other._record_id
            and self._site_id == other._site_id
            and self._user_id == other._user_id
            and self._group_id == other._group_id
            and self._components == other._components
            and self._start_time == other._start_time
            and self._stop_time == other._stop_time
        )

    def with_start_time(self, start_time: str) -> Record:
        self._start_time = start_time
        return self

    def with_stop_time(self, stop_time: str) -> Record:
        self._stop_time = stop_time
        return self

    #  def with_runtime(self, runtime: str) -> Record:
    #      self._runtime = runtime
    #      return self

    def record_id(self) -> str:
        return self._record_id

    def site_id(self) -> str:
        return self._site_id

    def as_dict(self) -> dict:
        return {
            "record_id": self._record_id,
            "site_id": self._site_id,
            "user_id": self._user_id,
            "group_id": self._group_id,
            "components": self._components.get(),
            "start_time": self._start_time,
            "stop_time": self._stop_time,
        }

    def as_json(self) -> str:
        return json.dumps(self.as_dict())

File: src/domain/record.rs
//! Record related types used for deserializing HTTP requests and serializing HTTP responses.

use super::{Component, ComponentTest, ValidName};
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};

#[derive(Serialize, Deserialize, Clone, Debug)]
pub struct RecordAdd {
    pub record_id: ValidName,
    pub site_id: ValidName,
    pub user_id: ValidName,
    pub group_id: ValidName,
    pub components: Vec<Component>,
    pub start_time: DateTime<Utc>,
    pub stop_time: Option<DateTime<Utc>>,
}

#[derive(Serialize, Deserialize, Clone, Debug)]
pub struct RecordUpdate {
    pub record_id: ValidName,
    pub site_id: ValidName,
    pub user_id: ValidName,
    pub group_id: ValidName,
    pub components: Vec<Component>,
    pub start_time: Option<DateTime<Utc>>,
    pub stop_time: DateTime<Utc>,
}

#[derive(Serialize, Deserialize, Clone, Debug)]
pub struct Record {
    pub record_id: String,
    pub site_id: Option<String>,
    pub user_id: Option<String>,
    pub group_id: Option<String>,
    pub components: Option<Vec<Component>>,
    pub start_time: DateTime<Utc>,
    pub stop_time: Option<DateTime<Utc>>,
    pub runtime: Option<i64>,
}

#[derive(Serialize, Deserialize, Clone, Debug, Default)]
pub struct RecordTest {
    pub record_id: Option<String>,
    pub site_id: Option<String>,
    pub user_id: Option<String>,
    pub group_id: Option<String>,
    pub components: Option<Vec<ComponentTest>>,
    pub start_time: Option<DateTime<Utc>>,
    pub stop_time: Option<DateTime<Utc>>,
}

impl RecordTest {
    pub fn new() -> Self {
        RecordTest::default()
    }

    pub fn with_record_id<T: AsRef<str>>(mut self, record_id: T) -> Self {
        self.record_id = Some(record_id.as_ref().to_string());
        self
    }

    pub fn with_site_id<T: AsRef<str>>(mut self, site_id: T) -> Self {
        self.site_id = Some(site_id.as_ref().to_string());
        self
    }

    pub fn with_user_id<T: AsRef<str>>(mut self, user_id: T) -> Self {
        self.user_id = Some(user_id.as_ref().to_string());
        self
    }

    pub fn with_group_id<T: AsRef<str>>(mut self, group_id: T) -> Self {
        self.group_id = Some(group_id.as_ref().to_string());
        self
    }

    pub fn with_component<T: AsRef<str>>(mut self, name: T, amount: i64, factor: f64) -> Self {
        if self.components.is_none() {
            self.components = Some(vec![])
        }
        self.components.as_mut().unwrap().push(ComponentTest {
            name: Some(name.as_ref().to_string()),
            amount: Some(amount),
            factor: Some(factor),
        });
        self
    }

    pub fn with_start_time<T: AsRef<str>>(mut self, start_time: T) -> Self {
        self.start_time = Some(
            DateTime::parse_from_rfc3339(start_time.as_ref())
                .unwrap()
                .with_timezone(&Utc),
        );
        self
    }

    pub fn with_stop_time<T: AsRef<str>>(mut self, stop_time: T) -> Self {
        self.stop_time = Some(
            DateTime::parse_from_rfc3339(stop_time.as_ref())
                .unwrap()
                .with_timezone(&Utc),
        );
        self
    }
}
